### PR TITLE
feat: Authentication support

### PR DESCRIPTION
**Pull request by amunra**
# Change

Added optional authentication support using the [ring](https://github.com/briansmith/ring) library.

This change closes https://github.com/questdb/c-questdb-client/issues/1.

To create keys, follow the QuestDB [authentication documentation](https://questdb.io/docs/reference/api/ilp/authenticate/).

# New APIs

## C API

```c
line_sender_sec_opts sec_opts;
memset(&sec_opts, 0, sizeof(sec_opts));
sec_opts.auth_key_id = "testUser1";
sec_opts.auth_priv_key = "5UjEMuA0Pj5pjK8a-fa24dyIf-Es5mYny3oE_Wmus48";
sec_opts.auth_pub_key_x = "fLKYEaoEb9lrn3nkwLDA-M_xnuFOdSt9y0Z7_vWSHLU";
sec_opts.auth_pub_key_y = "Dt5tbS1dEDMSYfym3fgMv0B99szno-dFc1rYF9t0aac";

line_sender_error* err = NULL;
line_sender* sender = line_sender_connect_secure("0.0.0.0", host, port, &sec_opts, &err);
```

Also see `examples/line_sender_c_example_auth.c`

## C++ API

```cpp
questdb::ilp::sec_opts sec_opts{
    "testUser1",  // auth_key_id
    "5UjEMuA0Pj5pjK8a-fa24dyIf-Es5mYny3oE_Wmus48",  // auth_priv_key
    "fLKYEaoEb9lrn3nkwLDA-M_xnuFOdSt9y0Z7_vWSHLU",  // auth_pub_key_x
    "Dt5tbS1dEDMSYfym3fgMv0B99szno-dFc1rYF9t0aac"};  // auth_pub_key_y
questdb::ilp::line_sender sender{host, port, sec_opts};
```

Also see `examples/line_sender_cpp_example_auth.cpp`

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -258,31 +280,170 @@ pub struct LineSender {
...
+        Ok(())
+    }
+
+    fn read_challenge(&mut self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        self.sock.set_read_timeout(Some(Duration::from_secs(15)))
+            .map_err(|io_err| map_io_to_socket_err(
+                "Failed to set read timeout on socket: ", io_err))?;
+        let mut reader = BufReader::new(&mut self.sock);
+        reader.read_until(b'\n', &mut buf)
```
Comment (eugenels) [inline src/lib.rs:407]: Why not just `read_line` directly into `self.output` ?

File: src/lib.rs [hunk 2]
```diff
@@ -258,31 +280,170 @@ pub struct LineSender {
...
-                let host_port = format!("{}:{}", host, port);
+                let host_port = format!("{}:{}", self.host, self.port);
                 let prefix = format!("Could not connect to {:?}: ", host_port);
                 map_io_to_socket_err(&prefix, io_err)
             })?;
-        Ok(Self {
+        let mut sender = LineSender {
             sock: sock,
             state: State::Connected,
             output: String::with_capacity(65536),
```
Comment (eugenels) [inline src/lib.rs:350]: If you already have a builder, have you considered making buffer capacity configurable?

File: src/lib.rs [hunk 3]
```diff
@@ -258,31 +280,170 @@ pub struct LineSender {
...
+        let mut buf = Vec::new();
+        self.sock.set_read_timeout(Some(Duration::from_secs(15)))
+            .map_err(|io_err| map_io_to_socket_err(
+                "Failed to set read timeout on socket: ", io_err))?;
+        let mut reader = BufReader::new(&mut self.sock);
+        reader.read_until(b'\n', &mut buf)
+            .map_err(|io_err| map_io_to_socket_err(
+                "Failed to read authentication challenge (timed out?): ",
+                io_err))?;
+        if buf.last().map(|c| *c).unwrap_or(b'\0') != b'\n' {
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/lib.rs:411]: No. It isn't: There could have been a closed socket, as such the data returned may be partial.

File: src/lib.rs (hunk 2 above)
Comment (amunra) [reply src/lib.rs]: Added.

File: src/lib.rs [hunk 4]
```diff
@@ -258,31 +280,186 @@ pub struct LineSender {
...
+    }
+
+    /// Connect synchronously.
+    pub fn connect(self) -> Result<LineSender> {
+        let addr: SockAddr = gai::resolve_host_port(self.host, self.port)?;
         let sock = Socket::new(Domain::IPV4, Type::STREAM, Some(Protocol::TCP))
-            .map_err(|io_err| map_io_to_socket_err("Could not open TCP socket: ", io_err))?;
+            .map_err(|io_err| map_io_to_socket_err(
+                "Could not open TCP socket: ", io_err))?;
         sock.set_nodelay(true)
```
Comment (puzpuzpuz) [inline src/lib.rs:345]: nit: is there a particular reason to disable naggling in ILP clients? I'm just curious.

File: src/ffi.rs [hunk 5]
```diff
@@ -283,6 +311,93 @@ pub extern "C" fn line_sender_connect(
...
+    }
+
+    let auth_key_id = unsafe { (*sec_opts).auth_key_id };
+    let auth_priv_key = unsafe { (*sec_opts).auth_priv_key };
+    let auth_pub_key_x = unsafe { (*sec_opts).auth_pub_key_x };
+    let auth_pub_key_y = unsafe { (*sec_opts).auth_pub_key_y };
+
+    if auth_key_id.is_null() && auth_priv_key.is_null() &&
+       auth_pub_key_x.is_null() && auth_pub_key_y.is_null() {
+        return true;
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/ffi.rs:339]: This is by design:
If a user wants to configure the authentication to be optional, they don't need to call different functions to create the connection, they can just call `line_sender_connect_secure`.
The previous `line_sender_connect` is there to avoid breaking the existing API.
That said, I have an API break coming up anyways (where the `..._name` type wil be split into two types).
Maybe this second function should just be simplified down to a single `line_sender_connect` function.

File: src/lib.rs (hunk 4 above)
Comment (amunra) [reply src/lib.rs:345]: Not sure I follow you: What do you mean by "disable nagging"?

You might have read `map_err` as consuming the error and throwing it away. That is not the case:
A bit of Rust background on error handling and an explaination of what's happening here:
* Functions in Rust return things that can go wrong as `Result<Thing, SomeError>`.
* This is a sum type that should be read roughly as "Either a `Thing` or a `SomeError` error.".
* My Rust API returns a single consistent error type.
* Here `.map_err` takes a function that can transform the error whilst keeping the value untouched.
* In other words, `.map_err` transforms any error and does not throw it away, i.e. maps `Result<T, E1>` to `Result<T, E2>`.
* The `?` is a short-hand notation that inserts an `if` statement and early-returns in case of errors.
 In other words, if you had a `let x: Result<T, E> = ...`, the type of the expression `x?` is simply `T`.

File: system_test/fixture.py [hunk 6]
```diff
@@ -41,6 +41,11 @@
 import urllib.parse
 import urllib.error
 
+
```
[the remark this answers is not in the document]
Comment (amunra) [reply system_test/fixture.py:44]: Nope, I forgot to add these.
In actual fact, I should have additional tests to:
* Connect with invalid certs.
* Connect with valid but incorrect certs.
* Connect auth client to non-auth server.
* Connect non-auth client to auth server.

File: src/lib.rs [hunk 7]
```diff
@@ -258,31 +280,186 @@ pub struct LineSender {
...
+impl LineSender {
+    fn send_key_id(&mut self, key_id: &str) -> Result<()> {
+        write!(&mut self.sock, "{}\n", key_id)
+            .map_err(|io_err| map_io_to_socket_err("Failed to send key_id: ", io_err))?;
+        Ok(())
+    }
+
+    fn read_challenge(&mut self) -> Result<Vec<u8>> {
+        let mut buf = Vec::new();
+        self.sock.set_read_timeout(Some(Duration::from_secs(15)))
```
[the remark this answers is not in the document]
Comment (amunra) [reply src/lib.rs:419]: Thinking about this.. I'm not sure it's worth the dev effort right now.

File: src/ffi.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply src/ffi.rs:339]: > Maybe this second function should just be simplified down to a single `line_sender_connect` function.

Yeah, that would be nice.

File: system_test/fixture.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply system_test/fixture.py:44]: Yeah, that would be a nice addition.

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply src/lib.rs:419]: It's fine to keep it as is. I was just wondering why we set the timeout for reads only.

File: src/lib.rs (hunk 4 above)
Comment (amunra) [reply src/lib.rs:345]: Ah, sorry: I never heard Nagle's algorithm referred as naggling.

I thought you were commenting on the error handling! 

The reason why nodelay is set comes from conversations with Vlad where he argued it should be set. 

In short, if there's a pending line which is small, we don't want to delay it. 

Of there's a lot of buffered data, the nodelay does nothing, so there's no harm keeping it.

File: system_test/fixture.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (amunra) [reply system_test/fixture.py:44]: Done. Tests implemented.

File: src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (amunra) [reply src/lib.rs:419]: Done. No-op.

File: src/lib.rs (hunk 4 above)
Comment (amunra) [reply src/lib.rs:345]: Done. No-op.

Comment (puzpuzpuz) [reply src/lib.rs:345]: > Also note that without nodelay, flush wouldn't actually flush.

The main promise of `write()` system call is to copy the data into OS buffer. All cases when naggling was disabled I saw in OSS projects are related with small enough messages and, usually, per key/row operations. In our case, writes are done in batches, so I would be surprised if disabling Nagle's algorithm really changes anything. Nevertheless, let's keep it as is. I was just curious if there is a particular reason to disable it here.

**Review by puzpuzpuz — APPROVED** [review]
